pkg/resource/resource.go

```diff
@@ -222,6 +229,40 @@ type Applicator interface {
 	Apply(context.Context, client.Object, ...ApplyOption) error
 }
 
+type shouldRetryFunc func(error) bool
+
+// An ApplicatorWithRetry applies changes to an object, retrying on transient failures
+type ApplicatorWithRetry struct {
+	Applicator
+	shouldRetry shouldRetryFunc
+	backoff     wait.Backoff
+}
+
+// Apply invokes nested Applicator's Apply retrying on designated errors
+func (awr *ApplicatorWithRetry) Apply(ctx context.Context, c client.Object, opts ...ApplyOption) error {
+	return retry.OnError(awr.backoff, awr.shouldRetry, func() error {
+		return awr.Applicator.Apply(ctx, c, opts...)
+	})
+}
+
+// NewApplicatorWithRetry returns an ApplicatorWithRetry for the specified
+// applicator and with the specified retry function.
+//   If backoff is nil, then retry.DefaultRetry is used as the default.
+func NewApplicatorWithRetry(applicator Applicator, shouldRetry shouldRetryFunc, backoff *wait.Backoff) Applicator {
```

Comment:
```suggestion
func NewApplicatorWithRetry(applicator Applicator, shouldRetry shouldRetryFunc, backoff *wait.Backoff) ApplicatorWithRetry {
```

Since the interface is programatically enforced already, I usually return just the struct itself so that if at some point a modification on the struct is needed, users can do so. If they require only the interface, which is the case here, then that'd work as well.

Reply:
```suggestion
func NewApplicatorWithRetry(applicator Applicator, shouldRetry shouldRetryFunc, backoff wait.Backoff) Applicator {
```

nitpick: I'd feel fine about requiring a backoff to be supplied since that's the interface of the underlying `retry.OnError`

Reply:
Hi @muvaf,
Thank you for the comments.

>  If they require only the interface, which is the case here, then that'd work as well.
Yes, I would like to hide the `resource.applicatorWithRetry` and only expose a constructor method with it. That's why the signature uses the interface type. I have also un-exported the type itself.

Reply:
> nitpick: I'd feel fine about requiring a backoff to be supplied since that's the interface of the underlying retry.OnError

I agree. We could get rid of providing a default for the back-off configuration. Why I preferred to use a default value is as follows:
The `retry` package provides two defaults back-off configurations with different semantics. The first one (`retyr.DefaultRetry`) is suggested to be used in contexts where there are competing modifiers for a resource (i.e., during resolution of HTTP-409 errors). The other one (`retry.DefaultBackoff`) is suggested to be used in contexts with unrelated changes. Because I expect, in general, a `resource.APIPatchingApplicator` to be wrapped by the `applicatorWithRetry`, I preferred to have the clients of `applicatorWithRetry` use `retry.DefaultRetry` as the default. 

But I don't have a strong opinion here, we can also remove a default value.

Reply:
> Yes, I would like to hide the resource.applicatorWithRetry and only expose a constructor method with it.

Why though? I'm thinking of the case where a consumer of `NewApplicatorWithRetry` gets an `Applicator`. If they'd like to change configuration after initialization or make use of new features we introduce to that struct that's not covered by the interface, they have to do a type casting. But if we return exported `ApplicatorWithRetry`, interface would still be satisfied and they'd get to work with the real object.

Reply:
Regarding the default backoff, I think user supplying `nil` and getting a default could be a bit misleading. In this context, `nil` is not a valid option so they might figure that out easily. But for defaulting, we usually accept [options](https://github.com/crossplane/crossplane-runtime/blob/f2418dc/pkg/reconciler/managed/reconciler.go#L386) for configuration and let the default overridden by the options if the user wants to. So it'd be consistent to expose options here as well if you'd like to have a default, but if you don't want to (since it's a single config parameter), requiring it to be supplied wouldn't be much of a hassle for users anyway IMO.

Reply:
@muvaf believes that we may utilize the applicatorWithRetry type in core crossplane repos. So he proposes to expose it. 
Agreed.